smarts/core/smarts.py

```diff
@@ -478,6 +478,11 @@ def attach_sensors_to_vehicles(self, agent_spec, vehicle_ids):
     def observe_from(self, vehicle_ids):
         return self._agent_manager.observe_from(self, vehicle_ids)
 
+    def history_set_start_elapsed_time(self, start_time: float):
```

Comment:
I am hesitant about this change since it specifically applies to the traffic history provider. Ideally I would want this to be a general `move_to_time()`.

Reply:
Liam's original PR on this issue tried to do that actually [on_this_line](https://github.com/huawei-noah/SMARTS/pull/163/files#diff-52e8b6282acc13c9a94300f00440b8a71ccf60c8c810becb246659aa297aa40aR287)
But JVillella mentioned: 
```
I don't think this is the right design. The reset(..., start_time=...) isn't actually doing what the API makes it seem. One would think it's offsetting the sim by the given start_time, but really just changing the internal clock.

This looks to be a one-off solution for the history data replay work and not useful to anything else (or applicable to anything else). Can we just offset the timestamps of the data replay data directly / modify the data replay input?
```
This seems to saying that move_to_time isn't useful to anything else besides history replay. The conclusion from the original PR was to:
1. the history provider needs to know the offset
2. mission starts at time 0 sec
which is done on this PR

Reply:
Another difficulty with setting the `self._elapsed_sim_time` is that it has to be done in `def reset` like the original PR. 

Reasons:
The goal is to reset self._elapsed_sim_time to be the time when the agent occurs. There are two hypothetical options to do this and avoid doing it in `def reset`:
1: we call `move_to_time()` after `reset()`, but this doesn't work since `reset` will keep stepping until the agent appears. Calling `move_to_time()` will just set elapsed_sim_time to the same value. There is no performance gain from doing this since reset still stepped through all timestamps before agent occurs
2: we call `move_to_time()` before `reset()`. This doesn't work either because `reset` will set `_elapsed_sim_time` to 0.

Reply:
I see that, I guess my concern here is that it is ballooning the API. It seems to be the wrong level to affect a specific provider from SMARTS instance. For instance I can say that the `SumoTrafficSimulation` also could use this time skip.

Reply:
I agree that this isn't a good level to affect a specific provider. I noticed we have the following apis in smarts.py
```
    def add_provider(self, provider):
        self._providers.append(provider)
    @property
    def providers(self):
        # TODO: Add check to ensure that action spaces are disjoint between providers
        # TODO: It's inconsistent that pybullet is not here
        return self._providers
```
Perhaps `history_vehicles_replacement_for_imitation_learning.py` should get smarts's providers and loop through the list to find traffic_history_provider and set the time_offset that way?

Reply:
Yes, that was what I was thinking.

Reply:
Yes, that was what I also arrived at. Ideally, though, we can have a better interface for getting the providers like `get_provider(<type>)`. This would be somewhat in line with something like the ECS architecture.

Reply:
Sounds good! i will do that